File: baekJoon/baekjoon_8972.cpp
// 미친 아두이노

#include <iostream>
#include <vector>
#include <string>
#include <string.h>
#include <math.h>
#include <limits.h>

using namespace std;

int r, c;
char arr[101][101];
int visited[101][101];
string command;

int dy[10] = { 0, 1, 1, 1, 0, 0, 0, -1, -1, -1 };
int dx[10] = { 0, -1, 0, 1, -1, 0, 1, -1, 0, 1 };

vector<pair<bool, pair<int, int>>> v;

int main() {
	cin.sync_with_stdio(0);
	cin.tie(0);
	int y, x;
	cin >> r >> c;
	for (int i = 0; i < r; i++) {
		cin >> arr[i];
		for (int j = 0; j < c; j++) {
			if (arr[i][j] == 'I') {
				y = i;
				x = j;
			}
			else if (arr[i][j] == 'R') v.push_back(make_pair(true, make_pair(i, j)));
		}
	}
	cin >> command;
	for (int X = 0; X < command.size(); X++) {
		arr[y][x] = '.';
		y += dy[command[X] - '0'];
		x += dx[command[X] - '0'];
		if (arr[y][x] == 'R') {
			cout << "kraj " << X + 1 << '\n';
			return 0;
		}
		arr[y][x] = 'I';

		for (auto& node : v) {
			if (!node.first) continue;
			int ny = node.second.first;
			int nx = node.second.second;
			arr[ny][nx] = '.';
			int min_val = INT_MAX;
			int d;
			for (int i = 1; i <= 9; i++) {
				int nny = ny + dy[i];
				int nnx = nx + dx[i];
				if (abs(y - nny) + abs(x - nnx) < min_val) {
					min_val = abs(y - nny) + abs(x - nnx);
					d = i;
				}
			}
			ny += dy[d];
			nx += dx[d];
			if (arr[ny][nx] == 'I') {
				cout << "kraj " << X + 1 << '\n';
				return 0;
			}
			node.second.first = ny;
			node.second.second = nx;
			visited[ny][nx]++;
		}

		for (auto& node : v) {
			if (!node.first) continue;
			if (visited[node.second.first][node.second.second] >= 2) node.first = false;
			else arr[node.second.first][node.second.second] = 'R';
		}
		memset(visited, 0, sizeof(visited));
	}
	for (int i = 0; i < r; i++)
		cout << arr[i] << '\n';
	return 0;
}

File: baekJoon/baekjoon_8972.py
import sys
input = sys.stdin.readline

if __name__ == '__main__':
    n, m = map(int, input().split())
    arr = [list(input().rstrip('\n')) for _ in range(n)]

    y, x = -1, -1
    v = []
    for i in range(n):
        for j in range(m):
            if arr[i][j] == 'I':
                y, x = i, j
            elif arr[i][j] == 'R':
                v.append([True, i, j])

    command = list(input().rstrip('\n'))
    for i in range(len(command)):
        command[i] = ord(command[i]) - ord('0')

    dy = [0, 1, 1, 1, 0, 0, 0, -1, -1, -1]
    dx = [0, -1, 0, 1, -1, 0, 1, -1, 0, 1]
    for i, d in enumerate(command):
        arr[y][x] = '.'
        y += dy[d]
        x += dx[d]
        if (arr[y][x] == 'R'):
            print("kraj {0}".format(i + 1))
            exit(0)
        arr[y][x] = 'I'

        visited = [[0] * m for _ in range(n)]
        for idx, node in enumerate(v):
            flag = node[0]
            ny = node[1]
            nx = node[2]
            if flag == False: continue
            min_val = float('inf')
            nd = 0
            for j in range(1, 10):
                nny = ny + dy[j]
                nnx = nx + dx[j]
                dis = abs(y - nny) + abs(x - nnx)
                if dis < min_val:
                    min_val = dis;
                    nd = j
            arr[ny][nx] = '.'
            ny += dy[nd]
            nx += dx[nd]
            if arr[ny][nx] == 'I':
                print("kraj {0}".format(i + 1))
                exit(0)
            v[idx][1] = ny
            v[idx][2] = nx
            visited[ny][nx] += 1

        for j, node in enumerate(v):
            flag = node[0]
            if flag == False: continue
            ny = node[1]
            nx = node[2]
            if visited[ny][nx] >= 2: v[j][0] = False
            else: arr[ny][nx] = 'R'

    for i in arr:
        print(''.join(map(str, i)))
